rosco-manifests/src/main/java/com/netflix/spinnaker/rosco/manifests/kustomize/KustomizationFileReader.java

```diff
@@ -0,0 +1,102 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.rosco.manifests.kustomize;
+
+import com.google.common.collect.ImmutableList;
+import com.netflix.spinnaker.kork.artifacts.model.Artifact;
+import com.netflix.spinnaker.kork.core.RetrySupport;
+import com.netflix.spinnaker.rosco.manifests.kustomize.mapping.Kustomization;
+import com.netflix.spinnaker.rosco.services.ClouddriverService;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.*;
+import java.util.stream.Collectors;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Component;
+import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.constructor.Constructor;
+import org.yaml.snakeyaml.representer.Representer;
+import retrofit.client.Response;
+
+@Component
+@Slf4j
+public class KustomizationFileReader {
+  private final ClouddriverService clouddriverService;
+  private final RetrySupport retrySupport = new RetrySupport();
+  private static final List<String> KUSTOMIZATION_FILENAMES =
+      ImmutableList.of("kustomization.yaml", "kustomization.yml", "kustomization");
+
+  public KustomizationFileReader(ClouddriverService clouddriverService) {
+    this.clouddriverService = clouddriverService;
+  }
+
+  public Kustomization getKustomization(Artifact artifact, String possibleName) {
+    Path artifactPath = Paths.get(artifact.getReference());
+    // sort list of names, trying the possibleName first.
+    List<String> names =
+        KUSTOMIZATION_FILENAMES.stream()
+            .sorted(
+                (a, b) ->
+                    a.equals(possibleName) ? -1 : (b.equals(possibleName) ? 1 : a.compareTo(b)))
+            .collect(Collectors.toList());
+
+    Kustomization k = null;
+    for (String name : names) {
+      try {
+        Artifact testArtifact = artifactFromBase(artifact, artifactPath.resolve(name).toString());
+        k = convert(testArtifact);
+        k.setKustomizationFilename(name);
+        break;
+      } catch (Exception e) {
+        log.error(
+            "kustomization file cannot be found for {}", artifactPath.resolve(name).toString());
+        continue;
+      }
+    }
+
+    if (k == null) {
+      throw new IllegalArgumentException(
+          "Unable to find any kustomization file for " + artifact.getName());
+    }
+
+    return k;
+  }
+
+  private Artifact artifactFromBase(Artifact artifact, String path) {
+    return Artifact.builder()
+        .reference(path)
+        .artifactAccount(artifact.getArtifactAccount())
+        .type(artifact.getType())
+        .build();
+  }
+
+  private Kustomization convert(Artifact artifact) throws IOException {
+    // TODO(ethanfrogers): figure out how to use safe constructor here.
```

Comment:
I'm not an expert, but does the current form prevent deserialization code injection?  I think that from a security perspective we'll need to solve this before merging (if this version hasn't already solved it).

Reply:
I'm honestly not sure. I know `SafeConstructor` only lets you deserialize to Java primitives so using it here will cause an error.

Reply:
For reference - Yes, using a `new Constructor(Kustomization.class)` will prevent code injection according to [this](https://lgtm.com/blog/swagger_snakeyaml_CVE-2017-1000207_CVE-2017-1000208).